### DISCUSSION TITLE
Rollback after automatic Upgrade

### DISCUSSION THREAD

**Discussion by theAkito** [k3s-io/k3s#6504] [Q&A]
https://docs.k3s.io/upgrades/automated

I wonder, if there is a rollback functionality related to the automated upgrade process.

It would be nice, if the upgrade would be monitored/recorded in some way, so, if something fails on a node, for example, there would be an automatic, unsupervised rollback, which automatically kicks in.

Is there any rollback functionality, specifically regarding the automated upgrade process, in place?

**Comment by brandond** [comment]
No, there is not currently any way to rollback Kubernetes version upgrades. This is not even guaranteed to be supported by Kubernetes, as storage versions and other bits may be non-reversibly altered during the upgrade. The only guaranteed way to roll back is to manually switch back to the previous release, and restore an etcd snapshot taken before the upgrade.

**Reply by theAkito** [reply]
https://docs.k3s.io/backup-restore

This backup method talks about data. You just mentioned data, as well.

But what about everything else? Is the restoration of only the data enough to restore everything, exactly the same as before?

**Reply by brandond** [reply]
Backup/restore will put the datastore back to how it was at a point in time. Backing up other data like node filesystems, persistent volumes, and so on is outside the scope of what the control-plane and datastore is responsible for.

**Reply by theAkito** [reply]
Yes, I get the persistent volumes part, though will a data backup restore all YAMLs, Cluster configuration and restore everything precisely the same, from a Kubernetes perspective? Or will something be missing (except the node & filesystem area)?

The reason I'm asking, is because I am used to backups in complicated systems to back up "everything" and this means, they only backup the data, but without configuration, sort orders, specific settings & all minor configurations.
These "backups" just restore the data, but forget about everything else.

In short: would it be in any way noticable, the Cluster was restored from a backup (again, except the filesystem, etc.)?

**Reply by theAkito** [reply]
For example, will the backup restore all past events & failed pods? Will the backup restore other past records?

**Reply by brandond** [reply]
It will restore everything that you can use kubectl to see.

Note that events are perhaps a bad example, as they have a 1 hour TTL and are automatically deleted from the datastore when they expire.

**Reply by theAkito** [reply]
> It will restore everything that you can use kubectl to see.

That sounds good.

> Note that events are perhaps a bad example, as they have a 1 hour TTL and are automatically deleted from the datastore when they expire.

Indeed. Was trying to think of a good example, but couldn't come up with one. It's something you can write an essay about, when you have such an issue, but when you don't, you cannot think of an example for it.

____

> Fifth, both plans have the version field set to v1.22.11+k3s2. Alternatively, you can omit the version field and set the channel field to a URL that resolves to a release of K3s. This will cause the controller to monitor that URL and upgrade the cluster any time it resolves to a new release. This works well with the [release channels](https://docs.k3s.io/upgrades/manual#release-channels). Thus, you can configure your plans with the following channel to ensure your cluster is always automatically upgraded to the newest stable release of K3s:

##### [Source](https://docs.k3s.io/upgrades/automated#configure-plans)

I wonder, if that means, "this is a nice feature, use at your own risk", or if it means, that maintainers are confident, that such an automatic release system is safe enough to be actually used.

Is there a method of having such completely automated backups be safe for whatever reason or would it be just as risky as any other upgrade on a production system?

____

>     Note: Users can and should use Rancher to upgrade their K3s cluster if Rancher is managing it.
> 
>         If you choose to use Rancher to upgrade, the following steps below are taken care of for you.
>         If you choose not to use Rancher to upgrade, you must use the following steps below to do so.

##### [Source](https://docs.k3s.io/upgrades/automated#overview)

Does that mean, only the plans are needed or not even the plans are needed? I would assume, the plans are needed either way, however that note in the documentation makes it seem like, they are not needed.

____

How wise is it to upgrade from a `k3s` version like `1.20` to the most current version `1.25`?
For example, in [this issue comment](https://github.com/k3s-io/k3s/issues/6041#issuecomment-1230972665) you specifically remarked the old Traefik version.
Do the same expectations apply for the `k3s` version leap example?
